### PR TITLE
HA flows Part 16: Add auto reroute

### LINKED ISSUE #5238: NPE when rerouting an HA-flow with protected path
Steps to reproduce:

- using the standard test topology create an HA-flow with the request below
- make sure the HA-flow is UP (use GET ha-flows)
- execute a reroute operation of this flow (/ha-flows/ha-flow-1/reroute)
Actual result: 
```json
{
    "correlation_id": "6aa682b1-7778-467d-8cd6-7f8c7e78584d : none",
    "timestamp": 1686567883328,
    "error-type": "Internal service error",
    "error-message": "Could not reroute HA-flow"
}
```
expected result: the HA-flow is rerouted sucessfully.

There is an error in the logs: `Flow ha-flow-1 error - AllocateProtectedResourcesAction failed: null` with the stack trace:
```
{
  "file": "AllocateProtectedResourcesAction.java",
  "method": "allocate_aroundBody0",
  "line": 82,
  "exact": false,
  "location": "stormjar.jar",
  "class": "org.openkilda.wfm.topology.flowhs.fsm.haflow.reroute.actions.AllocateProtectedResourcesAction",
  "version": "?"
}
```

The request to create an HA-flow:
```json
{
    "ha_flow_id": "ha-flow-1",
    "status": "In progress",
    "shared_endpoint": {
        "switch_id": "00:00:00:00:00:00:00:02",
        "port_number": 23,
        "vlan_id": 0,
        "inner_vlan_id": 0
    },
    "maximum_bandwidth": 0,
    "path_computation_strategy": "COST",
    "encapsulation_type": "transit_vlan",
    "max_latency": 0,
    "max_latency_tier2": 0,
    "ignore_bandwidth": true,
    "periodic_pings": true,
    "pinned": false,
    "priority": 0,
    "strict_bandwidth": false,
    "description": "HA flow test description 2",
    "allocate_protected_path": true,
    "diverse_with_flows": [],
    "diverse_with_y_flows": [],
    "diverse_with_ha_flows": [],
    "sub_flows": [
        {
            "flow_id": "ha-flow-1-a",
            "endpoint": {
                "switch_id": "00:00:00:00:00:00:00:09",
                "port_number": 22,
                "vlan_id": 0,
                "inner_vlan_id": 0
            },
            "status": "In progress",
            "description": "the first end point",
            "time_create": "2023-06-09T12:08:36.616Z",
            "time_update": "2023-06-09T12:08:36.618Z"
        },
        {
            "flow_id": "ha-flow-1-b",
            "endpoint": {
                "switch_id": "00:00:00:00:00:00:00:09",
                "port_number": 21,
                "vlan_id": 0,
                "inner_vlan_id": 0
            },
            "status": "In progress",
            "description": "the second end point",
            "time_create": "2023-06-09T12:08:36.617Z",
            "time_update": "2023-06-09T12:08:36.618Z"
        }
    ],
    "time_create": "2023-06-09T12:08:36.614Z",
    "time_update": "2023-06-09T12:08:37.038Z"
}
```

### PR DESCRIPTION
**Pull request by niksv**
Related to #5061
Closes #5238

### CODE REVIEW

**Review by artHalMirantis — APPROVED** [review]
can be released with the current test scope

tasks to do:
https://github.com/telstra/open-kilda/issues/5231

**Review by pablomuri — CHANGES_REQUESTED** [review]

File: src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java [hunk 1]
```diff
@@ -465,25 +486,77 @@ public int hashCode() {
...
-            if (haFlowStatus == null) {
-                haFlowStatus = subFlowStatus;
-            } else if (haFlowStatus == FlowStatus.DOWN && subFlowStatus != FlowStatus.DOWN
-                    || haFlowStatus == FlowStatus.UP && subFlowStatus != FlowStatus.UP) {
-                haFlowStatus = FlowStatus.DEGRADED;
+        }
+    }
+
+    /**
+     * Recalculates and set statuses of HA-sub fLows based on paths statuses.
```
Comment (Fedichk) [inline src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java:528]: fLows -> flows

**Review by Fedichk — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-messaging/src/main/java/org/openkilda/messaging/command/BaseRerouteRequest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.fasterxml.jackson.databind.PropertyNamingStrategy.SnakeCaseStrategy;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+
+import java.util.Set;
+
+@Data()
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-messaging/src/main/java/org/openkilda/messaging/command/BaseRerouteRequest.java:30]: looks like no

File: src-java/flowhs-topology/flowhs-messaging/src/main/java/org/openkilda/messaging/command/yflow/YFlowRerouteRequest.java [hunk 3]
```diff
@@ -15,42 +15,48 @@
...
+    @JsonProperty("y_flow_id")
     String yFlowId;
-    Set<IslEndpoint> affectedIsl;
-    boolean force;  // TODO: remove
-    String reason;
     boolean ignoreBandwidth;
 
     public YFlowRerouteRequest(@NonNull String flowId, String reason) {
-        this(flowId, Collections.emptySet(), reason);
+        this(flowId, new HashSet<>(), reason, false);
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-messaging/src/main/java/org/openkilda/messaging/command/yflow/YFlowRerouteRequest.java:44]: Don't know. Replaced back

File: src-java/flowhs-topology/flowhs-messaging/src/test/java/org/openkilda/messaging/SerializationTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/* Copyright 2020 Telstra Open Source
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-messaging/src/test/java/org/openkilda/messaging/SerializationTest.java:1]: done

File: src-java/flowhs-topology/flowhs-messaging/src/test/java/org/openkilda/messaging/SerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    private static final String REASON = "some reason";
+    private static final Set<IslEndpoint> AFFECTED_ISLS = Sets.newHashSet(
+            new IslEndpoint(new SwitchId(1), 1),
+            new IslEndpoint(new SwitchId(2), 2));
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private <T> T pass(T entity, Class<T> clazz) throws IOException {
+        return MAPPER.readValue(MAPPER.writeValueAsString(entity), clazz);
+    }
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-messaging/src/test/java/org/openkilda/messaging/SerializationTest.java:44]: done

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/create/actions/ResourcesAllocationAction.java [hunk 6]
```diff
@@ -156,6 +156,7 @@ private void createFlow(HaFlowRequest targetFlow) throws FlowNotFoundException,
             transactionManager.doInTransaction(() -> {
                 HaFlow haFlow = HaFlowMapper.INSTANCE.toHaFlow(targetFlow);
                 haFlow.setStatus(FlowStatus.IN_PROGRESS);
+                haFlow.getHaSubFlows().forEach(subFlow -> subFlow.setStatus(FlowStatus.IN_PROGRESS));
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/create/actions/ResourcesAllocationAction.java:159]: There are many cases when statuses may be different:
1. one subflow is broken:
 sub1: `down`
 sub2: `up`
 HA-flow `down`

2. one subflow is degraded
   sub1: `defraded`
   sub2: `up`
   HA-flow: `degraded`

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/delete/actions/ValidateHaFlowAction.java [hunk 7]
```diff
@@ -71,6 +71,7 @@ protected Optional<Message> performWithResponse(State from, State to, Event even
             // Keep it, just in case we have to revert it.
             stateMachine.setOriginalFlowStatus(haFlow.getStatus());
             haFlow.setStatus(FlowStatus.IN_PROGRESS);
+            haFlow.getHaSubFlows().forEach(subFlow -> subFlow.setStatus(FlowStatus.IN_PROGRESS));
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/delete/actions/ValidateHaFlowAction.java:74]: same

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/HaFlowValidateAction.java [hunk 8]
```diff
@@ -73,6 +73,7 @@ protected Optional<Message> performWithResponse(State from, State to, Event even
                         format("Could not swap paths: Protected path of HA-flow %s is not in ACTIVE state", haFlowId));
             }
             haFlow.setStatus(FlowStatus.IN_PROGRESS);
+            haFlow.getHaSubFlows().forEach(subFlow -> subFlow.setStatus(FlowStatus.IN_PROGRESS));
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/HaFlowValidateAction.java:76]: same

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/RecalculateFlowStatusAction.java [hunk 9]
```diff
@@ -70,6 +70,7 @@ protected void perform(State from, State to, Event event, HaFlowPathSwapContext
             if (status != haFlow.getStatus()) {
                 dashboardLogger.onHaFlowStatusUpdate(haFlowId, status);
                 haFlow.setStatus(status);
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/RecalculateFlowStatusAction.java:72]: what case? we are updating Subflow statuses in the next line

File: src-java/kilda-model/src/main/java/org/openkilda/model/FlowPath.java [hunk 10]
```diff
@@ -220,6 +220,17 @@ public String getHaFlowId() {
...
 
+    /**
+     * Returns HA-flow id, if path belongs to an HA-flow, null otherwise.
+     */
+    public HaFlow getHaFlow() {
+        HaFlowPath haFlowPath = getHaFlowPath();
+        if (haFlowPath != null) {
+            return haFlowPath.getHaFlow();
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/kilda-model/src/main/java/org/openkilda/model/FlowPath.java:231]: I think `if` is more readable in this case

File: src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java [hunk 11]
```diff
@@ -343,6 +345,25 @@ private Collection<PathId> getSubPathIds(Collection<HaFlowPath> haPaths) {
...
+        for (HaFlowPath haPath : getPaths()) {
+            if (haPath != null) {
+                for (FlowPath subPath : haPath.getSubPaths()) {
+                    if (subPath != null) {
+                        subPaths.add(subPath);
+                    }
+                }
+            }
+        }
+        return subPaths;
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java:364]: done

File: src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java (hunk 1 above)
Comment (niksv) [reply src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java:528]: good catch. fixed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/create/actions/ResourcesAllocationAction.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (pablomuri) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/create/actions/ResourcesAllocationAction.java:159]: ok, then it is ok

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/RecalculateFlowStatusAction.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (pablomuri) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/pathswap/actions/RecalculateFlowStatusAction.java:72]: yes, you're right

**Pull request merged**